test/replication-luatest/gh_8433_raft_is_candidate_test.lua

```diff
@@ -0,0 +1,106 @@
+            server.build_listen_uri('server2'),
+            server.build_listen_uri('server3'),
+        },
+    }
+
+    g.server1 = g.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = g.box_cfg,
+    }
+
+    -- Force server1 to be a leader for reliability and reproducibility
+    g.box_cfg.election_mode = 'voter'
+    g.server2 = g.replica_set:build_and_add_server{
+        alias = 'server2',
+        box_cfg = g.box_cfg,
+    }
+
+    g.proxy1 = proxy:new{
+        client_socket_path = server.build_listen_uri('server1_proxy'),
+        server_socket_path = server.build_listen_uri('server1'),
+    }
+
+    g.proxy2 = proxy:new{
+        client_socket_path = server.build_listen_uri('server2_proxy'),
+        server_socket_path = server.build_listen_uri('server2'),
+    }
+
+    luatest.assert(g.proxy1:start{force = true}, 'Proxy from 3 to 1 started')
+    luatest.assert(g.proxy2:start{force = true}, 'Proxy from 3 to 2 started')
+    g.box_cfg.replication[1] = server.build_listen_uri('server1_proxy')
+    g.box_cfg.replication[2] = server.build_listen_uri('server2_proxy')
+
+    g.server3 = g.replica_set:build_and_add_server{
+        alias = 'server3',
+        box_cfg = g.box_cfg,
+    }
+
+    g.replica_set:start()
+    g.replica_set:wait_for_fullmesh()
+end)
+
+g.after_all(function(g)
+    g.replica_set:drop()
+end)
+
+g.test_prevote_fail = function(g)
+    --
+    -- Test that applier failure doesn't start elections if the node
+    -- doesn't have enough quorum for doing that:
+    --     1. Fill leader_witness_map in order to not allow election_update_cb
+    --        start election on timeout.
+    --     2. Break applier to leader and wait until election pass
+    --        (4 * replication_timeout).
+    --     3. Break last applier and make sure, that election isn't started.
+    --
+    luatest.assert_equals(g.replica_set:get_leader(), g.server1)
+    local old_term = g.server1:get_election_term()
+    g.server3:exec(function()
+        box.cfg({election_mode = 'candidate'})
+    end)
+
+    -- We must be sure, that the server3 cannot start elections on timeout.
+    -- It's witness map should not be empty, when update_election_cb is executed
+    local death_timeout = g.box_cfg.replication_timeout * 4
+    luatest.helpers.retrying({timeout = death_timeout}, function()
+        if not g.server3:grep_log('leader is seen: true, state: follower') then
+            error("Witness map may still be empty")
+        end
+    end)
+
+    g.proxy1:pause()
+
+    -- Wait for the server3 to notice leader death and clear the
+    -- corresponding bit in leader_witness_map. The elections are not
+    -- supposed to start as the server2 says, that it can see the leader.
+    require('fiber').sleep(death_timeout)
+
+    g.proxy2:pause()
+
+    -- server3 should not start elections, as it doesn't
+    -- have enough quorum of healthy nodes to do that.
+    require('fiber').sleep(g.box_cfg.replication_timeout * 10)
```

Comment:
Again maybe we can wait until server3 notices upstream disconnection and then some (it might make the test more robust)

Reply:
And here's it enough to wait for applier death, as trigger will start elections immediately after it, without waiting for timeout